src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp

```diff
@@ -0,0 +1,184 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+#ifndef MIGRAPHX_GUARD_KERNELS_POOLING_HPP
+#define MIGRAPHX_GUARD_KERNELS_POOLING_HPP
+
+#include <migraphx/kernels/index.hpp>
+#include <migraphx/kernels/ops.hpp>
+#include <migraphx/kernels/math.hpp>
+#include <migraphx/kernels/array.hpp>
+
+namespace migraphx {
+
+struct max_pool
+{
+    MIGRAPHX_DEVICE_CONSTEXPR auto init() const { return lowest{}; }
+
+    template <class T, class U>
+    MIGRAPHX_DEVICE_CONSTEXPR auto operator()(T x, U y) const
+    {
+        return max(x, y);
+    }
+
+    template <class T>
+    MIGRAPHX_DEVICE_CONSTEXPR T final(T x, index_int) const
+    {
+        return x;
+    }
+};
+
+struct average_pool
+{
+    MIGRAPHX_DEVICE_CONSTEXPR auto init() const { return 0.0; }
+
+    template <class T, class U>
+    MIGRAPHX_DEVICE_CONSTEXPR auto operator()(T x, U y) const
+    {
+        return x + y;
+    }
+
+    template <class T>
+    MIGRAPHX_DEVICE_CONSTEXPR T final(T x, index_int y) const
+    {
+        return (y == 0) ? T{0.0} : T{x / y};
+    }
+};
+
+struct lpnorm_pool_base
+{
+};
+
+template <index_int P>
+struct lpnorm_pool : lpnorm_pool_base
+{
+    MIGRAPHX_DEVICE_CONSTEXPR auto init() const { return 0.0; }
+
+    template <class T>
+    MIGRAPHX_DEVICE_CONSTEXPR T apply(T x) const
+    {
+        if constexpr(P == 0)
+            return 1;
+        else if constexpr(P == 1)
+            return migraphx::abs(x);
+        else if constexpr(P == 2)
+            return x * x;
+        else
+            return migraphx::pow(migraphx::abs(x), T(P));
+    }
+
+    template <class T, class U>
+    MIGRAPHX_DEVICE_CONSTEXPR auto operator()(T x, U y) const
+    {
+        return x + apply(y);
+    }
+
+    template <class T>
+    MIGRAPHX_DEVICE_CONSTEXPR T final(T x, index_int) const
+    {
+        if constexpr(P == 0)
+            return 1;
+        else if constexpr(P == 1)
+            return x;
+        else if constexpr(P == 2)
+            return migraphx::sqrt(x);
+        else
+            return migraphx::pow(x, 1. / P);
```

Comment:
`1/P` would it be considered as constexpr ?

Reply:
What do you mean?

Reply:
Ah you mean can it be const-folded?

Reply:
Yes i mean if it can be const folded.

Reply:
Yea it should be const folded by the compiler.